opal/mca/btl/base/btl_base_am_rdma.c

```diff
@@ -0,0 +1,1064 @@
+{
+    int32_t result = 0;
+
+    switch (op) {
+    case MCA_BTL_ATOMIC_ADD:
+        result = opal_atomic_fetch_add_32 (addr, *operand);
+        break;
+    case MCA_BTL_ATOMIC_AND:
+        result = opal_atomic_fetch_and_32 (addr, *operand);
+        break;
+    case MCA_BTL_ATOMIC_OR:
+        result = opal_atomic_fetch_or_32 (addr, *operand);
+        break;
+    case MCA_BTL_ATOMIC_XOR:
+        result = opal_atomic_fetch_xor_32 (addr, *operand);
+        break;
+    case MCA_BTL_ATOMIC_SWAP:
+        result = opal_atomic_swap_32 (addr, *operand);
+        break;
+    case MCA_BTL_ATOMIC_MIN:
+        result = opal_atomic_fetch_min_32 (addr, *operand);
+        break;
+    case MCA_BTL_ATOMIC_MAX:
+        result = opal_atomic_fetch_max_32 (addr, *operand);
+        break;
+    default:
+        return OPAL_ERR_BAD_PARAM;
+    }
+
+    *operand = result;
+    return OPAL_SUCCESS;
+}
+
+static void mca_btl_base_am_rdma_response (mca_btl_base_module_t *btl, const mca_btl_base_receive_descriptor_t *desc) {
+    mca_btl_base_rdma_response_hdr_t *resp_hdr = (mca_btl_base_rdma_response_hdr_t *) desc->des_segments[0].seg_addr.pval;
+
+    assert (desc->des_segments[0].seg_len >= sizeof (*resp_hdr));
+
+    mca_btl_base_rdma_context_t *context = (mca_btl_base_rdma_context_t *)(uintptr_t) resp_hdr->context;
+
+    BTL_VERBOSE(("received response for RDMA operation. context=%p, size=%" PRIu64, context, resp_hdr->response_size));
+
+    if (MCA_BTL_BASE_AM_PUT != context->type) {
+        uint64_t local_address = resp_hdr->initiator_address ? resp_hdr->initiator_address : (uintptr_t) context->local_address;
+        if (local_address && MCA_BTL_BASE_AM_PUT != context->type) {
+            BTL_VERBOSE(("unpacking response from packet"));
+            /* if there is a result copy it out of the incoming buffer. if RDMA is being used (get/put or put/get)
+             * then the header should be the only thing in the incoming message. */
+            mca_btl_base_copy_from_segments (local_address, sizeof (*resp_hdr), desc->des_segments,
+                                             desc->des_segment_count);
+        }
+    }
+
+    if (context->total_size == opal_atomic_add_fetch_64 (&context->acknowledged, resp_hdr->response_size)) {
+        context->cbfunc (btl, desc->endpoint, context->local_address, context->local_handle,
+                         context->cbcontext, context->cbdata, OPAL_SUCCESS);
+        OBJ_RELEASE(context);
+    }
+}
+
+static void mca_btl_base_am_process_rdma (mca_btl_base_module_t *btl, const mca_btl_base_receive_descriptor_t *desc)
+{
+    /* not all btls work with these active message atomics. at this time
+     * all of the affected btls already have atomic support so there is
+     * no need to attempt to get the endpoint here. */
+    if (NULL == desc->endpoint) {
+        BTL_ERROR(("BTL is not compatible with active-message RDMA"));
+        abort();
```

Comment:
Can we be more graceful here?

Reply:
Resolving as this is the same as the discussion above.